### PR TITLE
Prefetch cleanup

### PR DESCRIPTION
**Pull request by steven-johnson**
This is (mostly) a cleanup pass to make the flow of Prefetch injection & lowering more obvious to the reader of the code (via commenting and minor code restructuring).

Notable exception: the HVX backend processing of Call::prefetch (and relevant runtime code) was refactored to make it (IMHO) less janky.

(Also some drive-by insertions of std::move where appropriate)

### CODE REVIEW

**Review by dsharletg — APPROVED** [review]

File: src/CodeGen_C.cpp [hunk 1]
```diff
@@ -2511,13 +2511,19 @@ void CodeGen_C::visit(const Call *op) {
                       " integer overflow for int32 and int64 is undefined behavior in"
                       " Halide.\n";
     } else if (op->is_intrinsic(Call::prefetch)) {
-        user_assert((op->args.size() == 4) && is_const_one(op->args[2]))
+        const Expr &base_address = op->args[0];
+        const Expr &base_offset = op->args[1];
+        const Expr &extent0 = op->args[2];  // unused
+        // const Expr &stride0 = op->args[3];  // unused
+
+        user_assert((op->args.size() == 4) && is_const_one(extent0))
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/CodeGen_C.cpp:2519]: Oops, sloppy work, thanks

File: src/CodeGen_LLVM.cpp [hunk 2]
```diff
@@ -3167,22 +3167,25 @@ void CodeGen_LLVM::visit(const Call *op) {
         llvm::CallInst *call = builder->CreateCall(base_fn->getFunctionType(), phi, call_args);
         value = call;
     } else if (op->is_intrinsic(Call::prefetch)) {
-        user_assert((op->args.size() == 4) && is_const_one(op->args[2]))
+        const Expr &base_address = op->args[0];
+        const Expr &base_offset = op->args[1];
+        const Expr &extent0 = op->args[2];
+        // const Expr &stride0 = op->args[3];  // unused
+        user_assert((op->args.size() == 4) && is_const_one(extent0))
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/CodeGen_LLVM.cpp:3174]: Ditto

File: src/Prefetch.cpp [hunk 3]
```diff
@@ -22,6 +22,22 @@ using std::set;
 using std::string;
 using std::vector;
 
+/**
+ * The steps by which prefetch directives are injected and lowered are a bit nonobvious;
+ * here's the overall flow at the time this comment was written:
+ *
+ *  - When the .prefetch() schedule directive is used, a PrefetchDirective is
+ *    added to the relevant Functiob
```
[the remark this answers is not in the document]
Comment (steven-johnson) [reply src/Prefetch.cpp:30]: Or maybe we just rename Function -> Functiob everywhere else?

**Comment by steven-johnson** [comment]
Failures are unrelated

**Pull request merged**